[REVIEW] Proposal for "proper" C/C++ API (issue #92)

jirikraus opened the pull request:
This is a proposal for a imporved C and C++ API. When reviewing this I suggest to start with `cuML/DEVELOPER_GUIDE.md` as that explains the used concepts.

teju85 commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->setDeviceAllocator()->allocate(n*sizeof(T), stream);
```
I believe you meant `getDeviceAllocator`?

jirikraus replied:
Yes I fixed it.

teju85 commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+{
+    T* temp_h = handle->setDeviceAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    handle->setDeviceAllocator()->deallocate(temp_h, n*sizeof(T), stream);
```
Same as above.

jirikraus replied:
Fixed.

teju85 commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+void foo( ML::cumlHandle* handle, .., cudaStream_t stream )
+{
+    ...
+    ML::device_buffer<T> temp( handle->getHostAllocator(), 0 )
```
Did you mean `getDeviceAllocator` here?

jirikraus replied:
Yes. I fixed it.

cjnolet commented:
Looks good, Jiri! I'm looking forward to using this.  

I am currently using cusparse in UMAP. It would be nice if the cusparse handle could be added to this API as well.

teju85 commented:
@cjnolet didn't realize that you were using cusparse. Can you add cusparse_wrappers, just like the cublas_wrappers, cusolver_wrappers, inside ml-prims, please?

jirikraus commented:
Thanks for the review @cjnolet I just added cusparseHandle_t to cumlHandle_impl.

teju85 commented on cuML/src/common/rmmAllocatorAdapter.hpp:
```diff
@@ -0,0 +1,81 @@
+
+#pragma once
+
+#include <rmm/rmm.h>
```
So, we expect rmm now to be installed before cuml build, right?

jirikraus replied:
No. This header is just provided as a convenience for users of cuML. It should not be used inside of cuML. E.g. if a cuser of cuML wants to use RMM for all device allocations code like this could be used:
 ```cpp
#include <cuML.hpp>
#include <rmm/rmm.h>
#include <common/rmmAllocatorAdapter.hpp>

void foo() {
    rmmOptions_t rmmOptions;
    rmmOptions.allocation_mode = PoolAllocation;
    rmmOptions.initial_pool_size = 0;
    rmmOptions.enable_logging = false;
    rmmError_t rmmStatus = rmmInitialize(&rmmOptions);
    if ( RMM_SUCCESS != rmmStatus )
        std::cerr<<"WARN: Could not initialize RMM: "<<rmmGetErrorString(rmmStatus)<<std::endl;

    std::unique_ptr<ML::cumlHandle> cumlHandle( new ML::cumlHandle() );

    std::shared_ptr<ML::rmmAllocatorAdapter> allocator( new ML::rmmAllocatorAdapter() );

    cumlHandle->setDeviceAllocator( allocator );

    cudaStream_t stream;
    cudaStreamCreate( &stream );
    cumlHandle->setStream( stream );

    cudaStreamDestroy( stream );
    if ( rmmIsInitialized( NULL ) )
        rmmFinalize();
}
```

dantegd replied:
@jirikraus thinking about it, could we put that `foo` function you just wrote (or something similar) somewhere as a convenience function for creating rmm enabled cumlHandle? From the python library I believe I will always call with RMM as the memory manager for the time being

teju85 replied:
Ah! Makes sense now. Thanks Jiri.

teju85 replied:
@dantegd maybe adding this as a code-block in the doxygen comment would be better for now?

dantegd replied:
Yeah, probably will end up adding it for the cython call, but in doxygen is fine for now

jirikraus replied:
@teju85 I agree this should go into a code-block in the doxygen comment. However before we concluded the discussion on the API I don't think it makes sense to invest much in documentation, which is the reason I have not added documentation yet.
@dantegd I think this code should be only in the documentation and examples. It needs to be written by the user of cuML as that is the one who knows if RMM is already initialized and how he want's to initialize. E.g. if cuML gets called from a python script which used cuDF to prepare data RMM was initialized already before. If we exluce RMM init and shutdown the function would only provide 
```cpp
std::shared_ptr<ML::rmmAllocatorAdapter> allocator( new ML::rmmAllocatorAdapter() );
cumlHandle->setDeviceAllocator( allocator );
```
which IMHO is not worth a separate helper function. Do you agree?

teju85 replied:
I agree with @jirikraus here. It is not worth such a helper function.

Sure Jiri. Makes sense to document this once we finalized the interfaces.

teju85 commented on cuML/src/cuML.hpp:
```diff
@@ -0,0 +1,55 @@
+
+namespace ML {
+
+class deviceAllocator {
```
Does it make sense to reuse the `DeviceAllocator` defined inside ml-prims here: https://github.com/rapidsai/cuml/blob/branch-0.6/ml-prims/src/device_allocator.h? Or, even upgrading this existing class' interface?

jirikraus replied:
I very much agree to this. This should not be duplicated. The advantage of the interface proposed in this pull request is that it allows asynchronous allocation and deallocation (if supported by the allocator).

teju85 replied:
I guess for now we can keep this duplication.
Once this PR gets through, we shall think about how to get rid of these duplications. (maybe we can just update the API of the ml-prims' `DeviceAllocator` itself?)

jirikraus replied:
I agree that we should do that after this is merged. I think the question how to avoid the duplication comes down to the question where the interface should live. Is my understanding correct that cuML has dependencies to ml-prims but not the other way around? In this case the Abstract base class definition of `ML::deviceAllocator` and `ML::hostAllocator` in `cuML.hpp` should simply be moved to a new hedaer in ml-prims that is then included from `cuML.hpp`.

teju85 replied:
Yes, your understanding is correct.
If you can take care of this change in this PR itself, that'd be great!

jirikraus replied:
Done I moved `deviceAllocator` and `hostAllocator` to `cuml/ml-prims/src/common/cuml_allocator.hpp`. Please resolve this discussion if you are OK with this in the scope of this PR.

teju85 commented on cuML/src/cuML.hpp:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <memory>
+#include <vector>
```
We don't need this header file, right?

jirikraus replied:
The header file is currently not used, but its intention is to contain the public external C++ interface of cuML.

teju85 replied:
Sorry, I meant we are not using `std::vector` in the current header file, so we don't this this '#include', right?

jirikraus replied:
Yes. I will remove it.

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
```
The comments are not updated for host_buffer?

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+ *     host_buffer<T> temp( handle->getHostAllocator(), 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
```
Is this a right example for host_buffer?

teju85 replied:
My bad. I missed that this is for cudaMallocHost use-case. Please ignore this comment.

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+        if ( n > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), 0 ));
+            CUDA_CHECK( cudaStreamSynchronize( 0 ) );
```
I'm a bit confused here. So effectively we are sync'ing on the default stream everytime we create a `device_buffer`?

jirikraus replied:
Yes, because there is no other stream available in the constructor to pass into the allocator and to synchronize on. I have chosen this approach because I think it would be confusing if a constructor is asynchronous. I would expect that an object is fully usable after construction. If a user provided stream should be used for the allocation and the synchronization should be avoided the technique described in the developer guide using `resize` and `release`  could be used. One alternative I could think is forcing users of device_buffer and host_buffer to use resize by not offering a constructor that takes a size. What do you think?

jrhemstad replied:
Yeah... you don't really want an asynchronous constructor. 

But I don't like that if I want asynchronous allocation/free I can't use RAII. 

How about an API that returns a `std::future` to an instance of a `device_buffer` allocated on a stream? This still allows for asynchronous memory allocation/free that still uses RAII.

teju85 replied:
Makes sense. Thanks for the clarification

jirikraus replied:
I get you point @jrhemstad regarding  `std::future`, however that does not change the issue of the constructor, right? It would replace the constructor with factory. Also you still have RAII when it is done like this
```cpp
template<typename T>
void foo( ML::cumlHandle* handle, .., cudaStream_t stream )
{
    ...
    ML::device_buffer<T> temp( handle->getDeviceAllocator(), 0 )
    
    temp.resize(n, stream);
    kernelA<<<grid, block, 0, stream>>>(..., temp.data(), ...);
    kernelB<<<grid, block, 0, stream>>>(..., temp.data(), ...);
    temp.release(stream);
}
```
in this case the pointer is still owned by temp. I.e. if the call to `release` does not happen for some reason (e.g. because a developer forgets to call it) the destructor of temp will make sure that the memory is released when temp goes out of scope. Does that make sense so you?

jirikraus replied:
I just pushed an update for device_buffer and host_buffer. Now the constructor takes a stream as an argument and the device_buffer and host_buffer are tied to that stream. The associated stream  a different stream can be changed by calling `resize`, `reserve´ or `release`.

teju85 replied:
Nice. This now looks very aligned with our idea of not having to use default stream in cuml or ml-prims!

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+    {
+        if ( 0 != _data ) 
+        {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), 0 );
```
Same as above comment for the ctor method. default stream sync is being imposed?

jirikraus replied:
See my reply there. I have choose this as the alternative with the least surprises for the user. If asynchronous free is required `release` that way it is transparent to the user of device_buffer that the buffer is freed asynchronously.

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+
+    ~device_buffer()
+    {
+        if ( 0 != _data ) 
```
minor knitpick: can you use `nullptr`, please?

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+            if ( _size > 0 ) {
+                CUDA_CHECK( cudaMemcpyAsync( new_data, _data, _size*sizeof(value_type), cudaMemcpyDeviceToDevice, stream ) );
+            }
+            if ( 0 != _data ) {
```
minor knitpick: can you use `nullptr`, please?

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+        if ( 0 != _data ) {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), stream );
+        }
+        _data = 0;
```
minor knitpick: can you use `nullptr`, please?

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+
+    ~host_buffer()
+    {
+        if ( 0 != _data ) 
```
minor knitpick: can you use `nullptr`, please?

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+        if ( _capacity > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), 0 ));
+            CUDA_CHECK( cudaStreamSynchronize( 0 ) );
```
same as above comments. default stream sync being imposed?

jirikraus replied:
Same reply as above :-)

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+    {
+        if ( 0 != _data ) 
+        {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), 0 );
```
same as above comments. default stream sync being imposed?

jirikraus replied:
Same reply as above :-)

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+            if ( _size > 0 ) {
+                CUDA_CHECK( cudaMemcpyAsync( new_data, _data, _size*sizeof(value_type), cudaMemcpyHostToHost, stream ) );
+            }
+            if ( 0 != _data ) {
```
minor knitpick: can you use `nullptr`, please?

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,165 @@
+        if ( 0 != _data ) {
+            _allocator->deallocate( _data, _capacity*sizeof(value_type), stream );
+        }
+        _data = 0;
```
minor knitpick: can you use `nullptr`, please?

jrhemstad commented on cuML/src/common/cumlHandle.cpp:
```diff
@@ -0,0 +1,418 @@
+
+void cumlHandle_impl::destroyResources()
+{
+    while ( !_cusparse_handles.empty() )
```
This is purely a aesthetic design critique, but a repeated pattern like this usually tells me there's a need for a more generic abstraction.

Perhaps a `for_each` over the elements of each vector and then `vector.clear()`.

jirikraus replied:
I agree to your statement. However given that I need to call a different destroy function for every vector I do not see how I would save a lot of code even using lambdas and `for_each`. However I am open for concrete suggestions :-)

jrhemstad replied:
> However I am open for concrete suggestions :- )

Fair enough! 

Okay, so... this is a bit more of an invasive change, but hear me out because I think it's beneficial for more than just solving this problem.

So for your vectors of resources, instead of maintaining a vector of the "raw" resource, you could create a simple wrapper class for each resource. For example, streams:

```
struct stream_wrapper{

   stream_wrapper(){
      CUDA_CHECK( cudaStreamCreate(&stream) );
   }

   ~stream_wrapper(){
      CUDA_CHECK( cudaStreamDestroy(&stream) );
   }

   // This allows using a `stream_wrapper` implicitly as if it were a cudaStream_t*,
   // i.e., the fact that it's a wrapper is invisible to the user  
   operator cudaStream_t*() { return &stream }; 

private:
  cudaStream_t stream;
};
```

And then... your vector of `cudaStream_t` would become a vector of `stream_wrapper`s...

```
std::vector<stream_wrapper>           _streams;
```

And then in `destroyResources` you can simply do `_streams.clear()` and let the wrapper's destructor handle all of the resource specific cleanup! 

You could then create similar wrappers for all of the other resources.

What do you think?

jirikraus replied:
I see how this would make the code safer. However it will also add quite some code. Since this is an implementation detail and we are primarily discussing the API here I would leave the destroy resources code as it currently is. Any objections?

jrhemstad replied:
> I see how this would make the code safer. However it will also add quite some code. Since this is an implementation detail and we are primarily discussing the API here I would leave the destroy resources code as it currently is. Any objections?

Absolutely just an implementation detail, don't let it block this PR.

teju85 replied:
@jrhemstad can you file a issue on this one so that we don't miss this, please?

jrhemstad commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+class device_buffer
+{
+public:
+    typedef std::size_t         size_type;
```
Nitpick: guidance these days is to use `using` instead of `typedef`. They're functionality equivalent, `using` is just a little more intuitive as it follows the pattern of variable assinment, e.g.,`using size_type = std::size_t` is akin to ` int x = y`.

jirikraus replied:
Fixed.

jrhemstad commented on cuML/src/common/cumlHandle.hpp:
```diff
@@ -0,0 +1,88 @@
+    std::vector<cublasHandle_t>         _cublas_handles;
+    std::vector<cusolverDnHandle_t>     _cusolverDn_handles;
+    std::vector<cusparseHandle_t>       _cusparse_handles;
+    std::shared_ptr<deviceAllocator>    _deviceAllocator;
```
I'm trying to understand the need for the allocators to be `shared_ptr`s instead of a `unique_ptr` member.

Shouldn't the `handle` be in charge of maintaining the lifetime of the allocator? That is, who else would have a reference to the allocator that requires using a `shared_ptr`?

jirikraus replied:
I have chosen a shared pointer as this allows the cuML user to reuse an allocator in other places in a save way. Does that make sense?

jrhemstad replied:
Yeah, I get that, but couldn't you also accomplish the same safety with a `unique_ptr`? 

My default is to always prefer a `unique_ptr` to a `shared_ptr` unless I absolutely need a `shared_ptr`, and I don't understand well enough yet why a `shared_ptr` is required here. 

My understanding is that there's a one-to-one relationship between an instance of a `ML::handle` and an instance of an allocator, and as such, the `handle` should have sole ownership over it's corresponding allocator. Is that understanding correct?

jirikraus replied:
From the point of view of anything that is implemented in cuML yes. However not from the point of view from a cuML user. If `setDeviceAllocator` would take a unique_ptr a cuML can't use a deviceAllocator that needs to outlive the cuML handle. Makes sense?

jrhemstad replied:
> If `setDeviceAllocator` would take a unique_ptr a cuML can't use a deviceAllocator that needs to outlive the cuML handle. Makes sense? Or am I missing something?

Ah yes, so if someone wants to reuse the same allocator between two cuML API calls, using a `unique_ptr` would be difficult.

Thanks for explaining it to me, I believe a `shared_ptr` is absolutely appropriate here.

jrhemstad commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,155 @@
+
+    device_buffer& operator=(const device_buffer& other) = delete;
+
+    device_buffer(std::shared_ptr<deviceAllocator> allocator, size_type n = 0)
```
It's a bit cumbersome to always have to pass an allocator as a parameter whenever I want to do an allocation with RAII.

It'd be nicer if it could be a template parameter such that you could have an alias for a `device_buffer` that always uses `my_allocator`.

jirikraus replied:
I have thought about that quite a bit. The advantage of passing in an allocator as a runtime argument over a template parameter are: 
* Irrespective of the used allocator the containers remain type compatible.
* We can easily have multiple different allocator objects, e.g. one per device one per thread, .... With a template argument we would need to reference a global object. 
Does that make sense to you?

teju85 commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getDeviceAllocator()->allocate(n*sizeof(T), stream);
```
@jrhemstad IIUC, page locked memory is what is being proposed with the host_allocator concept below.

teju85 replied:
Liked your suggestion about reducing verbosity.

jirikraus replied:
Interesting idea regarding the specialized allocators for Unified Memory and pinned host memory. However before we add these kind of things I would prefer to see the use case first. Currently I know that we need temporary host memory and temporary device memory. The interface is sufficient to support that and already gives the user of cuML quite some flexibility, e.g. he could pass in an allocator that allocates pinned host memory or one that allocates pageable host memory (current default is pinned host memory). Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.
Regarding the less verbose API: I actually had that in an earlier version I tried, but preferred a cleaner interface that does not duplicate functionality between classes. Also the `ML::device_buffer` and `ML::host_buffer` should be preferred for allocations IMHO. A verbose interface make the alternative less attractive. However I do not have a very strong opinion on this one so I am open to provide a device_allocate and host_allocate in cumlHandle.

jrhemstad replied:
> I would prefer to see the use case first.

I have no idea what the use-case may be :) This is my first time looking at cuML code. That said, we can never predict the future, so it's safest to assume that if something _can_ happen (like wanting to use UVM), it probably _will_ at some point and the design should be able to handle that.

> Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.

This would require the user defining their own custom allocator then? I think that's fine, just making sure I understand.

> Also the `ML::device_buffer` and `ML::host_buffer` should be preferred for allocations IMHO. A verbose interface make the alternative less attractive.

100% agreed that RAII constructs should be preferred. I had a sneaking suspicion that you left the `malloc/free`-like API verbose to discourage people from using it :)

jirikraus replied:
> > I would prefer to see the use case first.

> I have no idea what the use-case may be :) This is my first time looking at cuML code. That said,  we can never predict the future, so it's safest to assume that if something _can_ happen (like wanting to use UVM), it probably _will_ at some point and the design should be able to handle  that.

I would say the current design is flexible enough for that. Once the need comes up we can add allocation routines to the proposed allocator interfaces with default implementations, e.g. for manged memory calling through to `cudaMallocManaged()`. Since we won't need to make allocations with a new type of memory before that happens it won't require to touch old cuML code. And cuML users affected be the default implementation can easily override the default one. Does that seem reasonable to you?

> > Similar for temporary device memory he could use Unified Memory, pinned host memory or pinned device memory.

> This would require the user defining their own custom allocator then? I think that's fine, just making sure I understand.

Yes. A example for a custome allocator that falls back to is 
```cpp
class cachingDeviceAllocator : public ML::deviceAllocator {
public:
    cachingDeviceAllocator()
          , _allocator(8, 3, cub::CachingDeviceAllocator::INVALID_BIN, cub::CachingDeviceAllocator::INVALID_SIZE)
    {}

    virtual void* allocate( std::size_t n, cudaStream_t stream ) {
        void* ptr = 0;
        _allocator.DeviceAllocate( &ptr, n, stream );
        return ptr;
    }

    virtual void deallocate( void* p, std::size_t, cudaStream_t ) {
        _allocator.DeviceFree(p);
    }
    
private:
    cub::CachingDeviceAllocator _allocator;
};
```
That would be passed to cuML like this
```cpp
std::shared_ptr<ML::rmmAllocatorAdapter> allocator( new ML::rmmAllocatorAdapter() );
cumlHandle->setDeviceAllocator( allocator );
```

teju85 replied:
Let's just keep the current verbose API as it is. Depending on which one of these interfaces (device_buffer or allocate calls) folks end up preferring, we can then revisit the API restructuring.

teju85 commented:
@jirikraus I don't see the `nullptr` updates in the latest changes?

teju85 commented:
ah.. never mind my previous comment. The conversations still show up with the old code, whereas the commits themselves have your changes. I got confused by this distinction.

jirikraus commented:
I just added a new section to the proposed developer guide about Multi GPU Programming. Please take a look at it.

teju85 commented:
Multi GPU programming info LGTM.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+void foo( ML::cumlHandle* handle, .., cudaStream_t stream )
+{
+    ...
+    ML::device_buffer<T> temp( handle->getDeviceAllocator(), 0 )
```
The intention to call it buffer instead of vector is to make clear that this is a simple container enabling:
* asynchronous RAII
* Use of the proposed allocater interface
and **not** to offer the interface of a std::vector or thrust::device_vector. As that would duplicate quite some code and would come with a significant need for testing.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,93 @@
+    handle->getHostAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
````
I have thought about that quite a bit. The advantage of passing in an allocator as a runtime argument over a template parameter are:
* Irrespective of the used allocator the containers remain type compatible.
* We can easily have multiple different allocator objects, e.g. one per device one per thread, .... 

With a template argument we would need to reference a global object. Does that make sense to you?

jrhemstad replied:
All good points. 

I suppose since you are providing a host allocator and host_buffer, there's little reason for someone to use a `std::vector`. 

Another point in favor of "allocators as parameters" vs "template parameters" is that it allows stateful allocators, e.g., an allocator w/ a stream. This wouldn't be possible with an allocator as a template argument.

teju85 commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,99 @@
+```cpp
+void foo( ML::cumlHandle* handle, ...)
+{
+    cublasHandle_t cublasHandle = handle->getImpl()->getCublasHandle();
````
@jirikraus all of our cu*Handle objects typically support a `setStream` method and IIRC they all have default stream initialized in them? Maybe we should ask devs to prohibit from calling `setStream` on these handles?

teju85 replied:
Or maybe update our `cumlHandle_impl::setStream` to in turn also set streams for these handles appropriately?

jirikraus replied:
That is a good point. `setStream` is not really meant for cuML developers. It is meant for cuML users. Therefore I think it is better to pass around `const ML::cumlHandle&` in cuML which would prohibit changing the user provided stream. This also ties in with a comment from @jrhemstad regarding how to pass around the handle.

teju85 replied:
Sorry Jiri, if my previous comments confused you. I meant the following kind of usage pattern. In other words, should we allow developers to perform cublasSetStream on the internal cublas handle, explicitly, for example?
```
cublasHandle_t cublasHandle = handle->getImpl()->getCublasHandle();
cudaStream_t someStreamPickedByMe;
cublasSetStream(cublasHandle, someStreamPickedByMe);
```

teju85 replied:
On one hand, it sounds like a valid use-case if we want to allow devs to use internal streams and schedule cublas/cusolver calls on these streams.

On the other hand, these `cublasSetStream` kind of calls are persistent, easily causing false dependencies when the same cublasHandle gets called somewhere else in the pipeline!

jirikraus replied:
I also thought a bit more about this and one possibility would be that we drop the assumption that e.g. the `cublasHandle` managed by handle has any stream associated with it. Instead one should assume that it is required to set the stream before every cuBLAS call. For cuML and ml-prims developers I think everybody should use the wrapper is `cublas_wrappers.h` and we add a stream argument to the interface of those wrappers. E.g. for gemm
```cpp
template <typename T>
cublasStatus_t cublasgemm(cublasHandle_t handle, cublasOperation_t transA,
                          cublasOperation_t transB, int m, int n, int k,
                          const T *alfa, const T *A, int lda, const T *B,
                          int ldb, const T *beta, T *C, int ldc, cudaStream_t stream);
```
which would have the following specialization for dgemm
```cpp
template <>
inline cublasStatus_t
  cublasgemm(cublasHandle_t handle, cublasOperation_t transA,
             cublasOperation_t transB, int m, int n, int k, const double *alfa,
             const double *A, int lda, const double *B, int ldb,
             const double *beta, double *C, int ldc, cudaStream_t stream) {
  cublasStatus_t status = cublasSetStream(handle, stream);
  if ( CUBLAS_STATUS_SUCCESS != status )
    return status;
  return cublasDgemm(handle, transA, transB, m, n, k, alfa, A, lda, B, ldb,
                     beta, C, ldc);
}
```

Rational for this idea is that changing streams is s check operation and after an operation is en queued it is fine to change the stream. Only issue I see with this is potential races in a multi threaded enviroment, e.g. 
Thread A: sets stream to stream A on handle
Thread B: sets stream to stream B on handle
Thread B: launches Dgemm on handle in stream B
Thread B: launches Dgemm on handle in stream B (ups!)
We could either manage that by providing a cublasHandle_t wrapper containing a lock in addition to the handle that could be used in the cuBLAS wrappers. Or simply define a cuML, ml-prims developer rule that the wrapper are not thread safe and that one handle should be created per thread? The latter one would then require the `cumlHandle_t` to mange multiple cublasHandle's or that we disallow multi threading for cuML, ml-prims implementation. Considering the cuBLAS documentation having one handle per thread is propably what we should prefer:

[2.1.3. Thread Safety](https://docs.nvidia.com/cuda/cublas/index.html#thread-safety2)
The library is thread safe and its functions can be called from multiple host threads, even with the same handle. When multiple threads share the same handle, extreme care needs to be taken when the handle configuration is changed because that change will affect potentially subsequent CUBLAS calls in all threads. It is even more true for the destruction of the handle. So it is not recommended that multiple thread share the same CUBLAS handle.

What do you think?

teju85 replied:
My opinion (might be based, because I like the OPG approach!) is that we should disallow the usage of same cumlHandle across multiple threads.

Your cublas_wrappers idea is a much better one.

Implementation details:
If we allow `stream` as a default option in the above `cublasgemm` method you defined above, then I think this change can be made within the branch-0.6 timeframe itself. Else, it'll cause huge amount of changes across existing ml-prims and cuML code (not to mention the rework needed for ongoing PR's!).

teju85 replied:
Have filed issue #256 as our first step towards updating these *_wrapper header files. We shall update the existing methods to expose a stream as a default argument.

Will soon file another issue to remove the default-ness of stream argument across **all** ml-prims methods. We should not be launching work on default streams inside cuML or ml-prims!

jirikraus replied:
Agree. It makes sense that the implementation of cuML and ml-prims is single threaded in this case we won't have the issues I described. Just added two sentences in the Thread safety section of `DEVELOPER_GUIDE.md` to document it and put it up for discussion.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+To ensure that thrust algorithms are executed in the intended stream the `thrust::cuda::par` execution policy should be used (described in the section Device and Host memory allocations/Using thrust).
+
+# CUDA Resources
+Implementations of ML algorithms should not create reusable resources themselves. Instead they should use the existing one in `ML::cumlHandle_impl `. This allows to avoid constant creating and recreation of reusable resources such as CUDA streams, CUDA events or library handles. Please file a feature request in case a resource handle is missing in `ML::cumlHandle_impl `.
```
The basic idea is that the handle should own a pool of streams. So that we avoid constant creation and destruction of streams. I have not spend too much thought if this should just be a pool which can grow on demand with an interface like
```cpp
cudaStream_t stream = handle->getImpl()->getStream();
//use stream
handle->getImpl()->releaseStream( stream );
```
of if a handle simply has a fixed set of stream which are accessed like this
```cpp
cudaStream_t stream = handle->getImpl()->getStream(stream_id);
```
I think I would current prefer the former. The latter has the advantage that we would not need a releaseStream and a another smart RAII wrapper object that takes care that a stream is returned to the pool when going out of scope. What do you think?

jrhemstad replied:
> cudaStream_t stream = handle->getImpl()->getInternalStream(stream_id);

So there would need to be some way to query what the available `stream_id`s are, yes? 

> The fixed set is required to make `ML::detail::streamSyncer` work and over a stream pool

Does it need to be a fixed set? Couldn't you maintain a dynamically sized `vector` or `queue` of streams? It could grow as needed (but probably never shrink?) in response to additional requests for streams. 

`ML::detail::streamSyncer` could then iterate the streams in the pool at construction/destruction to synchronize accordingly. Maybe I'm missing something, but I think this should work even if the size of the pool isn't fixed.

jirikraus replied:
> > cudaStream_t stream = handle->getImpl()->getInternalStream(stream_id);
> 
> So there would need to be some way to query what the available `stream_id`s are, yes?

Yes that is needed.

> > The fixed set is required to make `ML::detail::streamSyncer` work and over a stream pool
> 
> Does it need to be a fixed set? Couldn't you maintain a dynamically sized `vector` or `queue` of streams? It could grow as needed (but probably never shrink?) in response to additional requests for streams.
> 
> `ML::detail::streamSyncer` could then iterate the streams in the pool at construction/destruction to synchronize accordingly. Maybe I'm missing something, but I think this should work even if the size of the pool isn't fixed.

What I had in mind is that stream created after the constructor of `ML::detail::streamSyncer` was called would not wait for the begin event recorded in the user stream. Would need to clarify this, but I guess stream creation implicitly synchronizing so this probably can't happen. However that is then the reason why we want to avoid dynamic stream creation. I just pushed a proposal adding `cudaStream_t getInternalStream( int sid ) const` and `int getNumInternalStreams() const` to `cumlHandle_impl` with a fixed number of internal streams.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getHostAllocator()->allocate(n*sizeof(T), stream);
```
See reply above :-)

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+void foo( ML::cumlHandle* handle, .. )
+{
+    ML::thrustAllocatorAdapter alloc( handle->getDeviceAllocator(), handle->getStream() );
+    auto execution_policy = thrust::cuda::par(alloc).on(handle->getStream());
```
Thanks for pointing this out. This is a great idea. I just tried it but to my understanding the implementation of `exec_policy` requires C++14 and I could not find a elegant way to express it with C++11 (which cuML is currently at). So I an implementation of `exec_policy` only conditionally and updated the developer guide mentioning the new possibilities with C++14. Let me know if I am missing something and we can make it work with C++11.

jrhemstad replied:
C++14 is only required for the `auto` return type. 

It's possible to support C++11 using trailing return type: https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.cbclx01/trailing_return.htm

It would just be a little more verbose, something like...
```
using T = decltype(thrust::cuda::par(rmm_allocator<char>{}));
inline exec_policy(cudaStream_t stream = 0) -> std::unique_ptr<T, std::function<void(T*)>
```

jirikraus replied:
Thanks @jrhemstad I could make it work with C++11 and just pushed the related updates.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,93 @@
+ensures the stream ordering behavior described above.
+
+## Using thrust
+To ensure that thrust algorithms are executed in the intended stream the `thrust::cuda::par` execution policy should be used (described in the section Device and Host memory allocations/Using thrust).
```
Fully agree to this. It is a good idea. However I think it requires C++14 and cuML is at C++11. See my comment above.

jirikraus commented:
Thanks @teju85 @jrhemstad for the thorough review and your good feedback. I currently have the following open items on my list:

* How to pass around `cumlHandle`? I currently think this should be as a const reference. I will work on a change.
* The name of `cumlHandle`. I agree that ML::cumlHandle is a dublication. I will update it to `ML::handle`.
* How to handle internal streams?

Please correct me if I am wrong, but I think I either addressed your other comments or there is an open discussion regarding them.

Thanks

Jiri

teju85 commented:
> passing around `cumlHandle`

Agreed. Better to pass around const-ref.

One other item is the `cumlHandle_impl`

I'm pretty sure we'd almost need all of the info inside the `cumlHandle_impl` object. (stream, cublas, cusolver and sooner cusparse handles). As such it also makes sense to migrate this impl class (please rename it to something meaningful) to inside of ml-prims itself. And since the dependency is from cuML to ml-prims (and should never be the other way around), this migration should be fine. What say?

jrhemstad commented:
@harrism should probably also review this for thoughts on allocator abstractions as well as for ideas about adopting something similar in cuDF.

jirikraus commented:
> One other item is the `cumlHandle_impl`
> 
> I'm pretty sure we'd almost need all of the info inside the `cumlHandle_impl` object (stream, cublas, cusolver and sooner cusparse handles) inside ml-prims. As such it also makes sense to migrate this impl class (please rename it to something meaningful) to inside of ml-prims itself. And since the dependency is from cuML to ml-prims (and should never be the other way around), this migration should be fine. What say?

I am not sure if cumlHandle_impl should be moved to ml-prims. I see that we want to avoid duplicating resources such as cublas or cusolver handles and that we also don't want to make the ml-prims API extremely verbose. However I see potential issues with sharing these resources, especially CUDA streams, between a cuML and ml-prims handle. As that could cause false dependencies. I therefore would suggest to introduce a new ml-prims handle and move the ownership of the allocators there. cumlHandle_impl will hold a ml-prim handle and reference the allocators in there. What do you think?

teju85 commented:
@jirikraus My main motivation behind writing that comment was to reduce duplication of resources (and thus code) between ml-prims and cuML. I'm ok with either full migration of the impl structure or a new handle in ml-prims with this impl struct being composed of it.

I think it's better addressed in a separate issue/PR.

teju85 commented:
JFYI, have filed issue #255 for the ml-prims handle thingy.

teju85 commented on ml-prims/src/common/cuml_allocator.hpp:
```diff
@@ -0,0 +1,35 @@
+/*
```
Knitpicks!
1. can you remove the 'cuml_' prefix from the filename, please?
2. Everything inside ml-prims is currently under `MLCommon` namespace. So, in the interest of keeping everything uniform, can you also update the namespace, please?

jirikraus replied:
I am open for other suggestions but I have added the `cuml_` prefix because the file needs to be part of the public interface as it is included by `cuML.hpp` for a binary build of the library it therefore needs to end up in the include directory. Since `allocator.hpp` is probably a very common name for an include file we need to avoid ambiguity here. We can either do putting it in a subfolder of an include path or by prefixing the file name. For the former we would need a concept how to manage include paths during development and after deployment. Do we have that in place? 
Regarding the namespace I have kept `ML` as I found other occurrences of `ML` in `ml-prims`. Thus asking for confirmation before I change it to `MLCommon`.

teju85 replied:
> re: namespace

Please go ahead with `MLCommon` itself. Later, I'll find those other occurrences and update them accordingly.

teju85 replied:
> cuml prefix

ah... that's a tricky one!
1. keeping this name kind of suggests that there's a dependency from ml-prims to cuml.
2. changing its name means we'll have trouble exposing it to public interface.

jirikraus replied:
OK. I changed the namespace. Regarding the filename, include path issue. I think this is a topic on its own mostly orthogonal to this PR. Are you OK if I file a issue to track it and we defer that discussion to a later time?

teju85 replied:
@dantegd do you have a better suggestion here for the filename?

If not, let's just keep the filename as it is and adjust it later as per the future needs.

teju85 commented on cuML/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,176 @@
+    void resize(const size_type new_size, cudaStream_t stream )
+    {
+        _stream = stream;
+        if ( _capacity < new_size )
```
knitpick: It sounds like we could replace this whole block (and the previous line) with a call to `reserve` method?

jirikraus replied:
Good spot. Just pushed a change streamlining the implementation.

teju85 commented on cuML/src/common/host_buffer.hpp:
```diff
@@ -0,0 +1,187 @@
+    void resize(const size_type new_size, cudaStream_t stream )
+    {
+        _stream = stream;
+        if ( _capacity < new_size )
```
similar comment on `reserve` usage here.

jirikraus replied:
Good spot. Just pushed a change streamlining the implementation.

teju85 commented on cuML/src/cuML.hpp:
```diff
@@ -0,0 +1,47 @@
+    std::shared_ptr<deviceAllocator> getDeviceAllocator() const;
+    void setHostAllocator( std::shared_ptr<hostAllocator> allocator );
+    std::shared_ptr<hostAllocator> getHostAllocator() const;
+    const cumlHandle_impl* getImpl() const;
```
If we are passing const-ref cumlHandle, then it also makes sense to have this method return const-ref, right?

jirikraus replied:
Yes. Fixed it.

teju85 commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,114 @@
+There are two simple container classes compatible with the allocator interface `ML::device_buffer` available in `cuML/src/common/device_buffer.hpp` and `ML::host_buffer` available in `cuML/src/common/host_buffer.hpp`. These allow to follow the [RAII idiom](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) to avoid resources leaks and enable exception safe code. These container ...
+```cpp
+template<typename T>
+void foo(const ML::cumlHandle_impl& h, ..., cudaStream_t stream )
````
I especially now like the idea about writing our algos as: `void cumlAlgo(const cumlHandle_impl& h, ...)`. This avoids a ton of `getImpl` calls in the downstream which would have been necessary, otherwise!

jirikraus replied:
Yup. Fixed it.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,93 @@
+# Asynchronous operations and stream ordering
+All ML algorithms should be as asynchronous as possible avoiding the use of the default stream (aka as NULL or `0` stream). If an implementation only requires a single CUDA Stream the stream from `ML::cumlHandle` should be used:
+```cpp
+void foo( ML::cumlHandle* handle, ...)
````
Switched to `const ML::cumlHandle_impl&`.

jirikraus commented:
Summary of latest updates:

* Switched to use `const ML::cumlHandle_impl& h` within cuML. `ML::cumlHandle` is for external use only.
* Added a proposal for internal streams: `cudaStream_t cumlHandle_impl::getInternalStream( int sid ) const` and `int cumlHandle_impl::getNumInternalStreams() const`
* Propose some content for the Thread safety section in `DEVELOPER_GUIDE.md`
* Moved `ML::device_buffer` to ml-prims `MLCommon::device_buffer`

The open question regarding the naming of `cumlHandle_impl` and `cumlHandle` is still open and I think more feedback on this is needed. I agree that `ML::cumlHandle` seems redudant but on the other hand `handle` is a pretty common name that potentially conflics with a variable name.

rietmann-nv commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,114 @@
+# Coding style
+
+# Error handling
+All calls to CUDA APIs should be done via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. Those macros take care of checking the return values of the used API calls and generate an exception in case the command was no successful. In case an exception needs to be avoided, e.g. when implementing a destructor, `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVE ...
```
"no successful" -> "not successful"

jirikraus replied:
Thanks. Fixed.

rlratzel commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,114 @@
+# cuML developer guide
```
For completeness, it seems like DEVELOPER_GUIDE.md and [CONTRIBUTING.md](https://github.com/rapidsai/cuml/blob/master/CONTRIBUTING.md) should cross reference each other somehow.

jirikraus replied:
Make sense. Just added a cross reference.

minseokl commented on ml-prims/src/common/device_buffer.hpp:
```diff
@@ -0,0 +1,164 @@
+ * @endcode
+ */
+template<typename T>
+class device_buffer
```
I am really enjoying reading your proposal. It is a cool stuff!
- One minor question: does it make sense to have a base class of `device_buffer` and `host_buffer` to reduce the duplicate code?

jirikraus replied:
Thanks for the feedback. That is a good suggestion. I will add a base class for `device_buffer` and `host_buffer` with the next iteration of this.

teju85 commented:
>The open question regarding the naming of cumlHandle_impl and cumlHandle is still open and I think more feedback on this is needed. I agree that ML::cumlHandle seems redudant but on the other hand handle is a pretty common name that potentially conflics with a variable name.

@jirikraus +1 for the slightly verbose `ML::cumlHandle` since it avoids any confusion while reading the code.

jirikraus commented:
Summary of latest updates:

* Added a section on C APIs to the developer guide proposal `DEVELOPER_GUIDE.md`
* Extended the proposed C API by missing functions (`cumlGetStream`, `cumlSetDeviceAllocator` and `cumlSetHostAllocator`)

Regarding the open question on the naming of `cumlHandle_impl` and `cumlHandle`: Please add a +1 to the comments below on the alternative you are in favor of. Conflict is that the currently proposed name `ML::cumlHandle` has redundant components but the alternative `ML::handle` is a pretty common name that potentially conflicts with a variable name.

jirikraus commented:
I am in favor of the name `ML::cumlHandle` to avoid conflicts with the pretty common name `handle`.

jirikraus commented:
I am in favor of the name `ML::handle` to avoid redundancy in the namespace variable name combination.

jirikraus commented:
Rebased to branch-0.7.

jirikraus commented:
Rebased back to branch-0.6 and changed target branch back to branch-0.6.

harrism commented on cuML/src/common/cumlHandle.cpp:
```diff
@@ -0,0 +1,435 @@
+        CUDA_CHECK( cudaStreamCreate(&stream) );
+
+        cublasHandle_t cublas_handle;
+        CUBLAS_CHECK( cublasCreate(&cublas_handle) );
```
Are you trying to avoid runtime overhead by pre-creating a handle to each LINALG library at startup? What is the cost of creation? What if the user doesn't ever use, say cuSolver? Does creating the handle cost anything in terms of memory usage?  Is it not acceptable to lazily create the handles, e.g. when the user first wants to use the library? (I know, implementation detail. Just food for thought, I don't need answers.)

jirikraus replied:
Yes the assumption is that usually with one handle one or multiple ML algorithms are called multiple times. So creating the library handle in `cumlHandle_impl` avoids creation and recreation. You have a good point regarding lazily creating the library handles as not all algorithms need all of them. On the other hand if they are created lazily a user of cuML does not have control when they are created. For that reason I am in favor of keeping it as it is and consider lazy initialization if a use case suffering for unnecessary initialization overhead arises. Does that make sense to you?

harrism replied:
Yup

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,119 @@
+Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
+
+## Thread safety
+cuML should be thread safe and its functions can be called from multiple host threads if they use different handles.
```
I fully agree the wording should be clear and "should" is not a good choice. Here I have started with "should" because this started as a proposal for discussion. However since that it is a proposal is obvious from the context.

jirikraus commented on cuML/src/common/allocatorAdapter.hpp:
```diff
@@ -0,0 +1,94 @@
+ * @Returns A Thrust execution policy that will use allocator for temporary memory
+ * allocation.
+ */
+inline auto exec_policy(std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream) -> std::unique_ptr<decltype(thrust::cuda::par(_decltypeHelper)),std::function<void(decltype(thrust::cuda::par(_decltypeHelper))*)> >
```
Nope cuML is still at C++11 :-(

harrism replied:
Any reason not to?

teju85 replied:
No particular reasons. Maybe this could be targeted for branch-0.7. Filed issue #304 for this one.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,119 @@
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
+Both the Host and the Device Allocators might allow asynchronous stream ordered allocation and deallocation. This can provide significant performance benefits so a stream always needs to be specified when allocating or deallocating (see [Asynchronous operations and stream ordering](# Asynchronous operations and stream ordering)).
+There are two simple container classes compatible with the allocator interface `MLCommon::device_buffer` available in `ml-prims/src/common/device_buffer.hpp` and `MLCommon::host_buffer` available in `ml-prims/src/common/host_buffer.hpp`. These allow to follow the [RAII idiom](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) to avoid resources leaks and enable exception safe c ...
````
Agree that should be clarified in the guide. I will add something.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,119 @@
+```cpp
+void foo(const ML::cumlHandle_impl& h, ... , cudaStream_t stream )
+{
+    auto execution_policy = ML::exec_policy(h.getDeviceAllocator(),stream);
````
Regarding passing in `ML::cumHandle_impl&`: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle. Besides that it avoids introducing a dependency of `ML::exec_policy` that is not really needed.
Regarding the one liner: I agree that this would be nicer. @jrhemstad said that this unfortunately requires thurst 1.9.4 or newer as the thrust execution policies with older thrust versions can't be initiated with temporaries. The current workaround with auto pointers also does not allow this because the auto pointer gets out of scope before the object is used. To my understanding another nice thing with thrust 1.9.4+ for this would be that one would not need to repeat the stream. We could provide a helper function that could be used like this:
```cpp
thrust::for_each(ML::exec_policy(alloc, stream), ... );
```

jirikraus commented on cuML/src/common/allocatorAdapter.hpp:
```diff
@@ -0,0 +1,94 @@
+    std::unique_ptr<T, decltype(deleter)> policy{new T(*alloc), deleter};
+    return policy;
+}
+
```
I see your point, but I am not convinced that we should add this function. Regarding passing in `const cumlHandle_impl& h` instead of `std::shared_ptr<deviceAllocator> allocator` does my argument that the execution policy does not depend on `cumlHandle_impl` made above convince you?

harrism replied:
I don't see an argument about this above.  Above what?

jirikraus replied:
Sorry I assumed you a reading top to bottom, propably github decided to hide part of the conversation. I was refering this reply made above:

Reply begin:

Regarding passing in ML::cumHandle_impl&: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle. Besides that it avoids introducing a dependency of ML::exec_policy that is not really needed.
Regarding the one liner: I agree that this would be nicer. @jrhemstad said that this unfortunately requires thurst 1.9.4 or newer as the thrust execution policies with older thrust versions can't be initiated with temporaries. The current workaround with auto pointers also does not allow this because the auto pointer gets out of scope before the object is used. To my understanding another nice thing with thrust 1.9.4+ for this would be that one would not need to repeat the stream. We could provide a helper function that could be used like this:

`thrust::for_each(ML::exec_policy(alloc, stream), ... );`

Reply end

jirikraus replied:
To be even more explicit I should have propably written: "Regarding passing in `ML::cumHandle_impl&` instead of `std::shared_ptr<deviceAllocator> allocator`: This is partly just my preference to be explicit, because the execution policy only needs the allocator and nothing else from handle"

jirikraus commented on cuML/src/common/cumlHandle.cpp:
```diff
@@ -0,0 +1,435 @@
+    return _userStream;
+}
+
+cudaStream_t cumlHandle_impl::getDeviceStream( int dev_idx ) const
```
Agree that is missing in the developer guide. I will add something for internal stream. 
I will drop the device stream. I added them to this PR as a proposal to support single process multi GPU for a transition period to the one process per GPU paradigm. But I think this is more confusing than helpful and it is probably better to migrate to one process per GPU for the affected algorithms directly.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
````diff
@@ -0,0 +1,119 @@
+    cudaStream_t stream = h.getStream();
+}
+```
+In case multiple streams are needed, e.g. to manage a pipeline, the internal streams available in `ML::cumlHandle_impl` should be used (see [CUDA Resources](# CUDA Resources)). If multiple streams are used all operations still need to be ordered according to `ML::cumlHandle::getStream()`, i.e. before any operation in any of the internal CUDA streams is started all previous work in `ML::cumlHandle ...
````
Thanks for the suggestions. I will add an example.

jirikraus commented on cuML/DEVELOPER_GUIDE.md:
```diff
@@ -0,0 +1,119 @@
+
+## Multi GPU
+ 
+The multi GPU paradigm of cuML is **O**ne **P**rocess per **G**PU (OPG). Each algorithm should be implemented in a way that it can run with a single GPU without any dependencies to any communication library. A multi GPU implementation can assume the following:;
```
Strange. For me this works when viewing with Chrome on github: https://github.com/jirikraus/cuml/blob/fea-ext-cuML-cpp-iface/cuML/DEVELOPER_GUIDE.md . What viewer are you using?

teju85 replied:
Seems to work for me in firefox too.

jirikraus commented:
@harrism thanks a lot for your thorough review and good feedback. I have incorporated almost all of you suggestions for `DEVELOPER_GUIDE.md` and hope that I did not miss to address any of your points. With the latest set of commit I changed the following:
- Added allocator adapter for STL containers. + a section in the `DEVELOPER_GUIDE.md` highlighting it.
- Dropped the multi GPU part of `cumlHandle_impl`
- Updated `DEVELOPER_GUIDE.md` based on suggestions from @harrism plus a few additions
  * Added a empty section about Testing and Unit Testing.
  * Added a section about Documentation stating that APIs should be documented with doxygen syntax.
- Added first version of doxygen documentation to the public interface (for internal interfaces like `cumlHandle_impl` this is still missing). I would prefer if we can merge this as it is and add missing documentation overtime in a collaborative effort. 

Regarding on remark to which I could not reply inline for some reason

> You don't have to reference a global argument with the latest Thrust device_vector -- the constructor has an allocator parameter. The same is true of std::vector:
> https://en.cppreference.com/w/cpp/container/vector/vector (see constructor #2 for example)
> 
> vector( size_type count, const T& value, const Allocator& alloc = Allocator());

As you say that would come with the dependency to the latest thrust version for device_vector. Regarding the motivation for device_buffer and host_buffer in general I have added something to the developer guide.

GPUtester commented:
Can one of the admins verify this patch?
